src/mono/mono/mini/method-to-ir.c

```diff
@@ -9898,8 +9902,16 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 			else {
 				klass = NULL;
 				field = mono_field_from_token_checked (image, token, &klass, generic_context, cfg->error);
-				if (!field)
-					CHECK_TYPELOAD (klass);
+				if (!field) {
+					if (cfg->compile_aot && (!klass || mono_class_has_failure (klass))) {
+						clear_cfg_error (cfg);
+						cfg->exception_type = MONO_EXCEPTION_NONE;
+						mono_emit_jit_icall (cfg, mono_throw_type_load, NULL);
+						break;
+					} else {
+						CHECK_TYPELOAD (klass);
+					}
+				}
```

Comment:
Can we have some helper function or macro that does the `if (cfg->compile_aot && COND) { /* emit throw */ }. else { CHECK_TYPELOAD(klass); }` logic?